gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/JobStateEventBuilder.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.metrics.event;
+
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+
+import org.apache.gobblin.metrics.GobblinTrackingEvent;
+
+
+/**
+ * An event reporting job state. It can be used to report the overall Gobblin job state or
+ * the state of an internal job, e.g MR job
+ */
+public class JobStateEventBuilder extends GobblinEventBuilder {
```

Comment:
what's special about this JobStateEvent so you need to invent a new class for doing this things? What are the alternative events achieving similar functionalities before ?

Reply:
The idea of all these specific event builders is to implement typed gobblin event. It is described in the class comment of `GobblinEventBuilder`. The motivation is that monitoring system built on gobblin events can inspect event attributes as java object fields, instead of relying on keys of metadata, which could be changed without control.

Currently, we are using the constants in `JobEvent` and `EventSubmitter` to send gobblin job state event.